File: backend/app/schemas/applications.py
"""Response shapes for the applications router.

Boundary models: the domain dataclasses (`app/domain/ranking.py`) and the stored
AI outputs (`app/ai/schemas.py`) stay snake_case and pure; these `*Out` models map
them to the camelCase wire. ``normalized`` and ``rawRow`` are intentionally
free-form dicts — their keys are raw form-field names (data like
``household_income``), not schema field names, so they pass through untouched.
"""

from typing import Any

from pydantic import Field

from app.schemas.base import RequestModel, ResponseModel
from app.schemas.openings import OpeningDetailsOut


class CommitteeOpeningOut(OpeningDetailsOut):
    phase: str


class HardFilterReason(ResponseModel):
    code: str
    message: str
    details: dict[str, Any] = {}


class Essay(ResponseModel):
    label: str
    question: str
    answer: str


class ScreeningFlagOut(ResponseModel):
    category: str
    summary: str
    evidence: str


class PetFactsOut(ResponseModel):
    """The pet inventory the screening AI extracted from the free-text pets field.
    Surfaced on the detail so the pet card can show the AI's ``reasoning`` — what it saw and
    how it classified it — at the finding, rather than making a member hunt the whole-pass
    narrative for the pet paragraph (extracting that slice would be brittle). The counts are
    what the deterministic pet rule judges. ``None`` when the app hasn't been screened."""

    dogs: int
    cats: int
    other_pets: list[str]
    reasoning: str


class DimensionContributionOut(ResponseModel):
    """Camel-cased view of the ranking ``DimensionContribution`` dataclass."""

    dimension_key: str
    name: str
    score: float
    weight: float
    impact: float
    confidence: str
    rationale: str
    evidence: str


class AIResultTraceOut(ResponseModel):
    """Provenance for one cached per-application model result. Cost and tokens describe
    its original generation allocation; a later run may reuse that result from cache."""

    model_id: str
    supports_reasoning_effort: bool
    reasoning_effort: str | None
    prompt_version: str
    input_tokens: int
    output_tokens: int
    cost_usd: float


class AIModelTraceOut(ResponseModel):
    """One exact model/reasoning pairing that contributed to a stored result."""

    model_id: str
    supports_reasoning_effort: bool
    reasoning_effort: str | None


class DimensionScoringTraceOut(ResponseModel):
    """Combined provenance for an applicant's current-dimension score results.

    A fresh scoring call commonly produces several dimension rows. On later updates,
    cached dimensions may originate from different calls, models, or prompt revisions,
    so this reports the exact total and all contributing provenance rather than
    pretending there was one call.
    """

    dimension_count: int
    models: list[AIModelTraceOut]
    prompt_versions: list[str]
    input_tokens: int
    output_tokens: int
    cost_usd: float


class ApplicationSummary(ResponseModel):
    id: int
    primary_email: str
    applicant_name: str | None = None
    co_applicant_name: str | None = None
    status: str
    status_source: str
    stale: bool
    hard_filter_reasons: list[HardFilterReason] = []
    child_count: int | None = None
    household_income: int | None = None
    # null = screening pass not run; int = flag count (0 = ran clean).
    flag_count: int | None = None
    flag_categories: list[str] | None = None
    # Whether the current member has starred (favourited) this applicant. Private
    # per member; a personal working aid with no effect on ranking or eligibility.
    starred_by_me: bool = False
    opening_ids: list[int] = []


class ApplicationDetail(ApplicationSummary):
    auto_status: str
    auto_status_source: str
    first_submitted_at: str | None = None
    last_submitted_at: str | None = None
    submission_version_count: int = 0
    normalized: dict[str, Any] | None = None
    essays: list[Essay] = []
    flags: list[ScreeningFlagOut] | None = None
    pet_facts: PetFactsOut | None = None
    raw_row: dict[str, Any] | None = None
    ai_narrative: str | None = None
    screening_trace: AIResultTraceOut | None = None
    dimension_scores: list[DimensionContributionOut] | None = None
    dimension_scoring_trace: DimensionScoringTraceOut | None = None
    # The current reviewer's private note. It is intentionally not part of the
    # application, source row, AI input, or any shared report.
    private_note: str = ""


class PrivateNoteUpdate(RequestModel):
    note: str = Field(max_length=10_000)


class ApplicationEnvelope(ResponseModel):
    """Single application is wrapped — it's an entity the SPA holds in state."""

    application: ApplicationDetail


class ApplicationListResponse(ResponseModel):
    # The whole pool, unpaginated; the client derives filters, sorting, and facet
    # counts (status/source/favourites) from it.
    applications: list[ApplicationSummary]
    openings: list[CommitteeOpeningOut]

File: frontend/src/types/applications.ts
import type { OpeningDetails } from "./core";
import type { DimensionContribution } from "./ranking";
import type { ReasoningEffort } from "./settings";

export type AppStatus = "eligible" | "ineligible";
export type StatusSource = "untouched" | "rules" | "ai" | "human";

// Which screening steps have run (persisted), so workflow gating survives a reload.
export type WorkflowState = {
  applicationsAvailable: boolean;
  screened: boolean;
  patternsDiscovered: boolean;
  candidatesScored: boolean;
  // Same truth the Rank no-op gate uses; the "needs re-run" badge reads this (not
  // score coverage), so a pool change still flags re-rank with full coverage.
  rankingCurrent: boolean;
};

// Per-AI-step coverage of the current scope. cached < inScope means results went
// stale, so the UI warns instead of a misleading done-check. Keys are absent for
// steps not yet computable (e.g. scoring before patterns exist).
export type Coverage = Partial<
  Record<"screened" | "candidatesScored", { cached: number; inScope: number }>
>;

export type AdminActions = {
  archivedOpeningsNeedingSelection: Array<{
    openingId: number;
    unitSizeBedrooms: number;
    moveInDate: string;
  }>;
  queuedEmailCount: number;
  quotaBlockedEmailCount: number;
  recentFailedEmailCount: number;
  oldestQueuedEmailAt: string | null;
  newestQueuedEmailAt: string | null;
  lastEmailAttemptAt: string | null;
};

export type EmailDeliveryIssue = {
  id: number;
  recipientEmail: string;
  messageKind: string;
  state: "queued" | "failed";
  attemptedAt: string;
  attemptCount: number;
  errorCode: string | null;
  quotaBlocked: boolean;
};

// Faceted counts: each facet reflects the other group's active filter, so the two
// filter groups stay consistent.
export type AppFacets = {
  status: Record<AppStatus, number>;
  source: Record<StatusSource, number>;
  // Count of the member's starred applications under the other active filters.
  favourites: number;
};

export type ApplicationSummary = {
  id: number;
  primaryEmail: string;
  applicantName: string | null;
  coApplicantName: string | null;
  status: AppStatus;
  statusSource: StatusSource;
  // True when machine findings changed since a human last reviewed.
  stale: boolean;
  hardFilterReasons: Array<{ code: string; message: string; details: Record<string, unknown> }>;
  childCount: number | null;
  householdIncome: number | null;
  // null = AI screening pass not run; int = flag count (0 = ran clean).
  flagCount: number | null;
  // Distinct flag categories from the latest pass (null if not run).
  flagCategories: string[] | null;
  // Whether the signed-in member has starred (favourited) this applicant. Private
  // per member; a personal working aid with no effect on ranking or eligibility.
  starredByMe: boolean;
  openingIds: number[];
};

export type CommitteeOpening = OpeningDetails & {
  phase: "upcoming" | "open" | "closed" | "archived";
};

export type Essay = {
  label: string;
  question: string;
  answer: string;
};

export type ScreeningFlag = {
  category: string;
  summary: string;
  evidence: string;
};

export type AIResultTrace = {
  modelId: string;
  supportsReasoningEffort: boolean;
  reasoningEffort: ReasoningEffort | null;
  promptVersion: string;
  inputTokens: number;
  outputTokens: number;
  costUsd: number;
};

export type DimensionScoringTrace = {
  dimensionCount: number;
  models: Array<{
    modelId: string;
    supportsReasoningEffort: boolean;
    reasoningEffort: ReasoningEffort | null;
  }>;
  promptVersions: string[];
  inputTokens: number;
  outputTokens: number;
  costUsd: number;
};

export type ApplicationDetail = ApplicationSummary & {
  // What the machine would decide from the current findings — i.e. the result of
  // clearing a human override. Lets the status control show the automatic verdict.
  autoStatus: AppStatus;
  autoStatusSource: StatusSource;
  firstSubmittedAt: string | null;
  lastSubmittedAt: string | null;
  submissionVersionCount: number;
  normalized: Record<string, unknown>;
  essays: Essay[];
  // null = screening pass not yet run for this application; [] = ran, clean.
  flags: ScreeningFlag[] | null;
  // The AI-extracted pet inventory and its interpretation. Null means no extraction is stored.
  petFacts?: { dogs: number; cats: number; otherPets: string[]; reasoning: string } | null;
  rawRow?: Record<string, unknown>;
  // The model's free-text reasoning from the latest screening pass.
  aiNarrative?: string | null;
  // Provenance for the latest screening result and current dimension score results.
  // Costs describe original generation allocations; results may be reused from cache.
  screeningTrace?: AIResultTrace | null;
  // This candidate's scores against the current run's dimensions, by |impact|
  // descending — the same ranking contributions the ranked-list row slices. null =
  // no run, or not scored under it.
  dimensionScores?: DimensionContribution[] | null;
  dimensionScoringTrace?: DimensionScoringTrace | null;
  // Private to the signed-in committee member; never included in AI inputs.
  privateNote: string;
};
